File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function distanceK(root: TreeNode | null, target: TreeNode | null, k: number): number[] {
    if (!root) return []

    const parentMap: Map<TreeNode, TreeNode | null> = new Map()

    let targetNode: TreeNode | null = null

    function dfs(node: TreeNode | null, parent: TreeNode | null) {
        if (!node) return

        parentMap.set(node, parent)
        
        if (node.val === target.val) {
            targetNode = node
        }

        dfs(node.left, node)
        dfs(node.right, node)
    }

    dfs(root, null)

    if (!targetNode) return []

    const queue: TreeNode[] = []
    const visited: Set<TreeNode> = new Set()

    queue.push(targetNode)
    visited.add(targetNode)

    let currentDistance: number = 0

    while (queue.length > 0) {
        const size: number = queue.length

        if (currentDistance === k) {
            return queue.map(node => node.val)
        }

        for (let i = 0; i < size; i++) {
            const currentNode: TreeNode = queue.shift()

            const neighbors: (TreeNode | null)[] = [currentNode.left, currentNode.right, parentMap.get(currentNode)]

            for (const neighbor of neighbors) {
                if (neighbor && !visited.has(neighbor)) {
                    visited.add(neighbor)
                    queue.push(neighbor)
                }
            }
        }

        currentDistance++
    }
    return []
};

File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        if not root:
            return []

        parent_map = {}

        def build_parent_map(node, parent=None):
            if node:
                parent_map[node] = parent
                build_parent_map(node.left, node)
                build_parent_map(node.right, node)

        build_parent_map(root)
        visited = {target}
        queue = deque([target])

        while queue:
            if k == 0:
                return [node.val for node in queue]

            for _ in range(len(queue)):
                curr = queue.popleft()

                for neighbor in (curr.left, curr.right, parent_map[curr]):
                    if neighbor and neighbor not in visited:
                        visited.add(neighbor)
                        queue.append(neighbor)

            k -= 1
        
        return []